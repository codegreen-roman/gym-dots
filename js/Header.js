import React from 'react'
import { connect } from 'react-redux'
import { shape, func, string } from 'prop-types'
import { setLocale } from 'react-redux-i18n'
import Text from './Text'

const _Header = props => (
    <section>
        <h4>I am a header</h4>
        <a href='#' onClick={() => props.changeLanguage('ru')}>changeLanguage</a>
        <p><Text text='greeting' /></p>
        <div>{props.exerciseSession.day}</div>
        <a href='#' onClick={props.pingSession}>ping</a>
    </section>
)

_Header.propTypes = {
    exerciseSession: shape({
        day: string
    }),
    pingSession: func,
    changeLanguage: func
}

const mapStateToProps = ({ exerciseSession }) => ({ exerciseSession })

const mapActionsToProps = dispatch => ({
    pingSession() {
        dispatch({ type: 'PING' })
    },
    changeLanguage(locale) {
        dispatch(setLocale(locale))
    }
})

export const Header = connect(mapStateToProps, mapActionsToProps)(_Header)
