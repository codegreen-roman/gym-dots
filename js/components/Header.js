import React from 'react'
import { connect } from 'react-redux'
import { shape, func, string } from 'prop-types'
import Text from './Text'
import ToggleMenu from './ToggleMenu'

const _Header = ({ exerciseSession, pingSession }) => (
    <section>
        <h4>I am a header</h4>
        <ToggleMenu />
        <p><Text text='greeting' /></p>
        <div>{exerciseSession.day}</div>
        <a href='#' onClick={pingSession}>ping</a>
    </section>
)

_Header.propTypes = {
    exerciseSession: shape({
        day: string
    }),
    pingSession: func
}

const mapStateToProps = ({ exerciseSession }) => ({ exerciseSession })

const mapActionsToProps = dispatch => ({
    pingSession() {
        dispatch({ type: 'PING' })
    }
})

export const Header = connect(mapStateToProps, mapActionsToProps)(_Header)
