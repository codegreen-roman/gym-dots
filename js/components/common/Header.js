import React from 'react'
import { connect } from 'react-redux'
import { shape, func, string } from 'prop-types'
import { Text } from '../Text'
import { ToggleMenu } from './ToggleMenu'
import { creators } from '../../state/actions'
import R from 'ramda'

const _Header = ({ exerciseSession, pingSession }) => (
    <section style={{ border: '2px solid #32CD32' }}>
        <h4>I am a header</h4>
        <ToggleMenu />
        <p><Text text='greeting' /></p>
        <div>{exerciseSession.day}</div>
        <a href='#' onClick={pingSession}>ping</a>
    </section>
)

_Header.propTypes = {
    exerciseSession: shape({
        day: string
    }),
    pingSession: func
}

const mapStateToProps = ({ exerciseSession }) => ({ exerciseSession })

const mapActionsToProps = dispatch => ({
    pingSession() {
        const app = R.compose(dispatch, creators.ping)
        app({})
    }
})

export const Header = connect(mapStateToProps, mapActionsToProps)(_Header)
