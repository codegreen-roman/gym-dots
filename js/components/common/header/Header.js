import React from 'react'
import { UserImage } from './UserImage'
import { string, object, func } from 'prop-types'
import { connect } from 'react-redux'
import { compose, propOr, isEmpty } from 'ramda'
import { authWith } from '../../../state/actions/firebase/databaseActions'

const getSafeNameOrEmptyString = propOr('', 'name')

const _Header = ({ dateStr, subTitle, loginWith, auth, exerciseName }) => {

    const rootStyle = {
        padding: '1rem',
        display: 'flex',
        justifyContent: 'space-between',
        borderBottom: '1px solid rgba(0, 0, 0, 0.15)',
        fontFamily: 'sans-serif'
    }

    const { user } = auth
    const renderTitle = () => isEmpty(exerciseName) ? subTitle : exerciseName

    const renderLoginOrUser = () => {

        const { user } = auth

        if (user) {
            return (
                <div className='username'>
                    <span>{user.displayName}</span>
                </div>
            )
        }

        return (
            <div>
                <button className='twitter' onClick={() => loginWith('twitter')}>login with twitter</button>
                <button className='facebook' onClick={() => loginWith('facebook')}>login with facebook</button>
            </div>
        )
    }

    return (
        <section style={rootStyle}>

            <div className='left-side' style={{
                display: 'flex',
                flexDirection: 'column',
                justifyContent: 'space-between'
            }}>
                <a className='h-menu' style={{ fontSize: '16px' }}>&#9776;</a>
                <div className='h-date' style={{ fontSize: '12px', marginLeft: '4rem' }}>
                    {dateStr}
                </div>
                <div className='h-title' data-test='currently' style={{ fontSize: '16px', marginLeft: '3rem' }}>
                    {renderTitle()}
                </div>
            </div>

            <div className='right-side'>
                {renderLoginOrUser()}
                <UserImage image={user && user.photoURL || undefined} />
            </div>

        </section>
    )
}

_Header.propTypes = {
    auth: object.isRequired,
    loginWith: func.isRequired,
    dateStr: string.isRequired,
    subTitle: string.isRequired,
    exerciseName: string.isRequired
}

_Header.defaultProps = {
    exerciseName: ''
}

const mapStateToProps = ({ auth, currentExercise }, { dateStr, subTitle }) => ({
    auth,
    dateStr,
    subTitle,
    exerciseName: getSafeNameOrEmptyString(currentExercise)
})

const mapActionsToProps = dispatch => ({
    loginWith: compose(dispatch, authWith)
})

export { _Header }
export const Header = connect(mapStateToProps, mapActionsToProps)(_Header)
