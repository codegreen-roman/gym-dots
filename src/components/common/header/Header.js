import React from 'react'
import { UserImage } from './UserImage'
import { string, func } from 'prop-types'
import { isEmpty } from 'ramda'
import {
    header,
    headerLeftSide,
    headerDate,
    headerText,
    headerMenu,
    userWrapper,
    userPic
} from './Header.glamor.js'
import { Icon } from '../icon/Icon'
import { branch, RenderNothing } from '@utils/helpers'

export const User = ({user, logout, photoURL}) => (
    <section {...userWrapper}>
        <span {...userPic} data-test='username'>{user}</span>
        <button className='logout' onClick={logout}>
            <UserImage image={photoURL} />
        </button>
    </section>
)

User.propTypes = {
    user: string,
    logout: func,
    photoURL: string
}


export const Header = ({ dateStr, subTitle, userDisplayName, exerciseName, logout, photoURL }) => {

    const renderTitle = () => isEmpty(exerciseName) ? subTitle : exerciseName

    const renderLogoutAndUser = () => branch(userDisplayName, <User user={userDisplayName} logout={logout} photoURL={photoURL} />, <RenderNothing />)

    return (
        <header {...header}>

            <div className='left-side' {...headerLeftSide}>
                <Icon
                    iconName='menu'
                    color='red'
                    viewBox='0 0 32 32'
                    width={16}
                    height={16}
                    {...headerMenu}
                />
                <div>
                    <div data-test='date' {...headerDate}>
                        {dateStr}
                    </div>
                    <div data-test='currently' {...headerText}>
                        {renderTitle()}
                    </div>
                </div>
            </div>

            <div>
                {renderLogoutAndUser()}
            </div>

        </header>
    )
}

Header.propTypes = {
    userDisplayName: string.isRequired,
    photoURL: string,
    logout: func.isRequired,
    dateStr: string.isRequired,
    subTitle: string.isRequired,
    exerciseName: string.isRequired
}

Header.defaultProps = {
    exerciseName: ''
}
