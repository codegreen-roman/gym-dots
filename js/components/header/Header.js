import React from 'react'
import { connect } from 'react-redux'
import { shape, func, string } from 'prop-types'
import { Text } from '../Text'
import { ToggleMenu } from '../common/ToggleMenu'
import { ping } from '../../state/actions/index'
import R from 'ramda'

const days = ['MONDAY', 'WEDENSDAY', 'FRIDAY']

const _Header = ({ exerciseSession: { day : currentDay }, pingSession }) => {

    const handleDayClick = (day) => {
        return () => pingSession(day)
    }

    const renderDayLink = day => {

        const style = {
            margin: '1rem',
            background: currentDay === day ? '#FF6347' : 'white'
        }
        return (
            <a key={day} href='#' style={style} onClick={handleDayClick(day)}>{day}</a>
        )
    }
    const renderDays = R.compose(R.map(renderDayLink))

    return (
        <section style={{ border: '2px solid #32CD32', margin: '2rem' }}>
            <h4>I am a header</h4>
            <ToggleMenu />
            <p><Text text='greeting' /></p>
            {renderDays(days)}
        </section>
    )
}

_Header.propTypes = {
    exerciseSession: shape({
        day: string
    }),
    pingSession: func
}

const mapStateToProps = ({ exerciseSession }) => ({ exerciseSession })

const mapActionsToProps = dispatch => ({
    pingSession(day) {
        const app = R.compose(dispatch, ping)
        app({ day })
    }
})

export { _Header }
export const Header = connect(mapStateToProps, mapActionsToProps)(_Header)
