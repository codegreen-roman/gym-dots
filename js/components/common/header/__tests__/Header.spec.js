import React from 'react'
import { Header } from '../Header'
import { shallow } from 'enzyme'
import { merge } from 'ramda'
import { shallowToJson } from 'enzyme-to-json'

const defaultProps = {
    dateStr: 'Saturday, Jul 15th',
    subTitle: 'Start Workout',
    exerciseName: '',
    auth: {},
    loginWith: () => {
    },
}

const setup = (props = defaultProps) => {
    return shallow(<Header {...props} />)
}

describe('Header component', () => {

    test('has section type of the root component', () => {
        const wrapper = setup()
        expect(wrapper.type()).toBe('section')
    })

    test('has 2 divs as 1st children', () => {
        const wrapper = setup()
        expect(wrapper.find('section > div').length).toBe(2)
    })

    describe('not logged in', () => {

        const props = {
            dateStr: 'Saturday, Jul 15th',
            subTitle: 'Start Workout',
            auth: {},
            loginWith: jest.fn()
        }

        let wrapper
        let twitterBtn
        let facebookBtn

        beforeEach(() => {
            wrapper = setup(props)
            twitterBtn = wrapper.find('button.twitter')
            facebookBtn = wrapper.find('button.facebook')
        })

        it('should call the loginWith with twitter provider', () => {
            twitterBtn.simulate('click')
            expect(props.loginWith).toBeCalledWith('twitter')
        })

        it('should call the loginWith with facebook provider', () => {
            facebookBtn.simulate('click')
            expect(props.loginWith).toBeCalledWith('facebook')
        })
    })

    describe('logged in', () => {

        const props = {
            dateStr: 'Saturday, Jul 15th',
            subTitle: 'Start Workout',
            auth: {
                user: {
                    displayName: 'Roman'
                }
            },
            loginWith: jest.fn()
        }

        let wrapper
        let twitterBtn
        let facebookBtn
        let usernameDiv

        beforeEach(() => {
            wrapper = setup(props)
            twitterBtn = wrapper.find('button.twitter')
            facebookBtn = wrapper.find('button.facebook')
            usernameDiv = wrapper.find('div.username')
        })

        it('should not have login buttons', () => {
            expect(twitterBtn.length).toBe(0)
            expect(facebookBtn.length).toBe(0)
        })

        it('should have a div with a class of username and text Roman', () => {
            expect(usernameDiv.length).toBe(1)
            expect(usernameDiv.text()).toBe('Roman')
        })
    })

    describe('Left side of the header', () => {

        let root
        let menuElement
        let dateElement
        let titleElement
        const currentDate = 'Saturday, Jul 15th'

        beforeAll(() => {
            root = setup().find('section > div.left-side')
            menuElement = root.find('a.h-menu')
            dateElement = root.find('div.h-date')
            titleElement = root.find('div.h-title')
        })

        test('has a menu icon wrapped in a tag', () => {
            expect(menuElement.length).toBe(1)
        })

        test('has a menu icon as a value', () => {
            expect(menuElement.text()).toBe('â˜°')
        })

        test('has dateElement with a date value Saturday, Jul 15th', () => {
            expect(dateElement.text()).toBe(currentDate)
        })

        test('has titleElement with a value Start Workout', () => {
            expect(titleElement.text()).toBe('Start Workout')
        })
    })

    describe('Exercise name exist', () => {

        const props = merge(defaultProps, { exerciseName: 'Pull Up' })
        let header
        let titleElement


        beforeAll(() => {
            header = setup(props)
            titleElement = header.find('[data-test="currently"]')
        })

        it('should match the snapshot', () => {
            expect(header).toMatchSnapshot()
        })

        it('should should show title with the exerciseName = Pull Up', () => {
            expect(titleElement.text()).toBe('Pull Up')
        })

    })

    describe('Shallow snapshot', () => {
        test('matches the previous', () => {
            const component = setup()
            const tree = shallowToJson(component)
            expect(tree).toMatchSnapshot()
        })
    })
})
