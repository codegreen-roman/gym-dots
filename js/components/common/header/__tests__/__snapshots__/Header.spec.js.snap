// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Exercise name exist should match the snapshot 1`] = `
<section
  style={
    Object {
      "borderBottom": "1px solid rgba(0, 0, 0, 0.15)",
      "display": "flex",
      "fontFamily": "sans-serif",
      "justifyContent": "space-between",
      "padding": "1rem",
    }
  }
>
  <div
    className="left-side"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
      }
    }
  >
    <a
      className="h-menu"
      style={
        Object {
          "fontSize": "16px",
        }
      }
    >
      ☰
    </a>
    <div
      className="h-date"
      style={
        Object {
          "fontSize": "12px",
          "marginLeft": "4rem",
        }
      }
    >
      Saturday, Jul 15th
    </div>
    <div
      className="h-title"
      data-test="currently"
      style={
        Object {
          "fontSize": "16px",
          "marginLeft": "3rem",
        }
      }
    >
      Pull Up
    </div>
  </div>
  <div
    className="right-side"
  >
    <div>
      <button
        className="twitter"
        onClick={[Function]}
      >
        login with twitter
      </button>
      <button
        className="facebook"
        onClick={[Function]}
      >
        login with facebook
      </button>
    </div>
    <UserImage
      image="https://s3.amazonaws.com/uifaces/faces/twitter/antjanus/128.jpg"
    />
  </div>
</section>
`;

exports[`Header component Shallow snapshot matches the previous 1`] = `
<section
  style={
    Object {
      "borderBottom": "1px solid rgba(0, 0, 0, 0.15)",
      "display": "flex",
      "fontFamily": "sans-serif",
      "justifyContent": "space-between",
      "padding": "1rem",
    }
  }
>
  <div
    className="left-side"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
      }
    }
  >
    <a
      className="h-menu"
      style={
        Object {
          "fontSize": "16px",
        }
      }
    >
      ☰
    </a>
    <div
      className="h-date"
      style={
        Object {
          "fontSize": "12px",
          "marginLeft": "4rem",
        }
      }
    >
      Saturday, Jul 15th
    </div>
    <div
      className="h-title"
      data-test="currently"
      style={
        Object {
          "fontSize": "16px",
          "marginLeft": "3rem",
        }
      }
    >
      Start Workout
    </div>
  </div>
  <div
    className="right-side"
  >
    <div>
      <button
        className="twitter"
        onClick={[Function]}
      >
        login with twitter
      </button>
      <button
        className="facebook"
        onClick={[Function]}
      >
        login with facebook
      </button>
    </div>
    <UserImage
      image="https://s3.amazonaws.com/uifaces/faces/twitter/antjanus/128.jpg"
    />
  </div>
</section>
`;
